### PR TITLE
fix millisecond fraction being handled with wrong scale

### LINKED ISSUE #61: Millisecond timestamps parsed incorrectly
The `DateTime::parse_str(...)` constructor supports various different formats, but millisecond-granularity timestamps are parsed incorrectly when they have a fractional part.

Testcase to reproduce:

```rust
use speedate::DateTime;

fn timestamp_ms(dt: DateTime) -> i64 {
    dt.timestamp() * 1000 + i64::from(dt.time.microsecond / 1000)
}

#[test]
fn from_fractional_float_str() {
    assert_eq!(
        timestamp_ms(DateTime::parse_str("1711445175471.865").unwrap()),
        1711445175471
    )
}
```

Failure:

```
---- from_fractional_float_str stdout ----
thread 'from_fractional_float_str' panicked at tests/datetime_timestamp_bytes.rs:9:5:
assertion `left == right` failed
  left: 1711445176335
 right: 1711445175471
```

Analysis: the difference between the two values is `864`, which is close to the fractional part `.865`. It seems that the fractional part of a float input is always interpreted on the seconds scale, even if the integral part of the input is interpreted on the milliseconds scale.

This is not a bug in the above `timestamp_ms()` helper. The same problem can be observed when accessing this functionality via Pydantic:

```python
import pydantic  # v2.6.4
from datetime import datetime

class Model(pydantic.BaseModel):
    x: datetime

Model(x="1711445175471.865").x.timestamp() * 1000
#=> 1711445176335.99
```

Originally I though this was a bug in the [custom float parsing](https://github.com/pydantic/speedate/blob/ecbe681c5a2ac84303adfc667ee982609de6ee98/src/numbers.rs#L58) which seems to accumulate rounding errors because it skips the checks in `f64::from_str()`, but I now think the issue is actually in the usage of the `DateTime::from_timestamp(ts, micros)` constructor by the `parse_bytes_with_config()` constructor:

https://github.com/pydantic/speedate/blob/ecbe681c5a2ac84303adfc667ee982609de6ee98/src/datetime.rs#L344-L346

I am not sure how to fix this. It is impossible to call `DateTime::from_timestamp(ts, micros)` with a non-zero value for `micros` without already knowing how the `ts` part will be interpreted.

### PR DESCRIPTION
**Pull request by davidhewitt**
Fixes #61

### CODE REVIEW

**Comment by davidhewitt** [comment]
I didn't try, but I wouldn't be concerned if it did. The behaviour is currently wrong and corrupting data 😬

**Comment by davidhewitt** [comment]
Well, I guess you could make the argument that even if the timestamp is a millisecond timestamp the fraction is always treated as seconds fraction, but I think that's just extremely non intuitive and likely to trip everyone up.

**Comment by sydney-runkle** [comment]
> Well, I guess you could make the argument that even if the timestamp is a millisecond timestamp the fraction is always treated as seconds fraction, but I think that's just extremely non intuitive and likely to trip everyone up.

Agreed, I think this fix makes sense.

File: src/date.rs [hunk 1]
```diff
@@ -206,9 +206,13 @@ impl Date {
...
     /// ```
     pub fn from_timestamp(timestamp: i64, require_exact: bool) -> Result<Self, ParseError> {
-        let (timestamp_second, _) = Self::timestamp_watershed(timestamp)?;
+        let (timestamp_second, millis) = Self::timestamp_watershed(timestamp)?;
         let d = Self::from_timestamp_calc(timestamp_second)?;
         if require_exact {
+            if millis != 0 {
+                return Err(ParseError::DateNotExact);
+            }
+
```
Comment (sydney-runkle) [inline src/date.rs:215]: Why do we need to raise an error here?

Comment (sydney-runkle) [reply src/date.rs:215]: Also, isn't the second returned value in `microseconds`?

Comment (sydney-runkle) [reply src/date.rs:215]: I guess maybe I don't understand the date not exact concept...

File: src/datetime.rs [hunk 2]
```diff
@@ -339,14 +339,36 @@ impl DateTime {
...
+                        let Some(fract_integers) = int_parse_bytes(fract) else {
+                            return Err(e);
+                        };
+                        let multiple = 10f64.powf(max_digits as f64 - fract.len() as f64);
+                        Self::from_timestamp_with_config(
+                            timestamp,
+                            // FIXME should we error if the fraction is too long?
+                            // We have TimeConfig truncate / error option.
+                            (fract_integers as f64 * multiple).round() as u32,
+                            config,
```
Comment (sydney-runkle) [inline src/datetime.rs:366]: I would say if we use a `max_digits` pattern above, it seems intuitive to then error if the number of decimal digits violates that standard.

**Review by sydney-runkle — COMMENTED** [review]
A few questions / comments, but looks great overall!

File: src/datetime.rs (hunk 2 above)
Comment (davidhewitt) [reply src/datetime.rs:366]: It's a behavioural change but I think correct given the spirit of the config.

File: src/date.rs (hunk 1 above)
Comment (davidhewitt) [reply src/date.rs:215]: Yes, it's in microseconds. I'll split into a separate PR as I think this is a correctness fix but sort of drive-by here.

Comment (sydney-runkle) [reply src/date.rs:215]: I don't mind the fix being here, but was just confused about the var name here

Comment (davidhewitt) [reply src/date.rs:215]: #67

File: src/datetime.rs [hunk 3]
```diff
@@ -339,14 +339,47 @@ impl DateTime {
...
+                    Some(fract) => {
+                        // fraction is either:
+                        // - up to 3 digits of millisecond fractions, i.e. microseconds
+                        // - or up to 6 digits of second fractions, i.e. milliseconds
+                        let max_digits = if timestamp > MS_WATERSHED { 3 } else { 6 };
+                        let Some(fract_integers) = int_parse_bytes(fract) else {
+                            return Err(e);
+                        };
+                        if config.microseconds_precision_overflow_behavior
+                            == MicrosecondsPrecisionOverflowBehavior::Error
```
Comment (sydney-runkle) [inline src/datetime.rs:364]: Where is this `MicrosecondsPrecisionOverflowBehavior` coming from? Do we need to add support for this upstream on `config`?

File: src/datetime.rs [hunk 4]
```diff
@@ -339,14 +339,47 @@ impl DateTime {
...
+                            && fract.len() > max_digits
+                        {
+                            return Err(if timestamp > MS_WATERSHED {
+                                ParseError::MillisecondFractionTooLong
+                            } else {
+                                ParseError::SecondFractionTooLong
+                            });
+                        }
+                        // Technically this is rounding
```
Comment (sydney-runkle) [inline src/datetime.rs:373]: Maybe add a note about this being in opposition to `truncate` which is what the setting is called? Is there any way to unify these (either truly truncate, or change the setting name)?

**Review by sydney-runkle — APPROVED** [review]
Nice, I feel particularly good about the new `MilisecondFractionTooLong` given the existence of `SecondFractionTooLong` already!

File: src/datetime.rs (hunk 3 above)
Comment (davidhewitt) [reply src/datetime.rs:364]: It already exists (and is implemented for the form `hh:mm:ss.1234567`)

**Comment by sydney-runkle** [comment]
Truncation note looks great! Thanks.

File: src/datetime.rs [hunk 5]
```diff
@@ -339,14 +339,50 @@ impl DateTime {
...
+                    // TODO might want to reconsider this later?
+                    Some(b"") | None => Self::from_timestamp_with_config(timestamp, 0, config),
+                    Some(fract) => {
+                        // fraction is either:
+                        // - up to 3 digits of millisecond fractions, i.e. microseconds
+                        // - or up to 6 digits of second fractions, i.e. milliseconds
+                        let max_digits = if timestamp > MS_WATERSHED { 3 } else { 6 };
+                        let Some(fract_integers) = int_parse_bytes(fract) else {
+                            return Err(e);
+                        };
```
Comment (latk) [inline src/datetime.rs:362]: Thank you for fixing my issue (#61). But I wonder about this strategy here, reusing an integer parsing function for a fractional part:

* does this handle leading zeros correctly? (I think yes, thanks to the $10^d$ scaling trick.)
* does this handle other aspects of integer notation correctly, e.g. signs? (I think not.)
* does this handle long fractions if the "precision overflow" check isn't used? (I think yes, thanks to the $10^d$ scaling trick.)

As an example, consider an input like `1711445175.+07186` which should probably be invalid, not parsed as if the `+` were a zero.

Perhaps the `i = int_parse_bytes(fract)` function could be copied as a `(i, magnitude) = fract_parse_bytes(fract)` function, where the `magnitude` would be safer than relying on `fract.len()`?

**Comment by sydney-runkle** [comment]
@latk,

See https://github.com/pydantic/speedate/pull/71 for a rework of this - I think most of your questions are addressed.

**Pull request merged**